iOS: Add share extension

sirnacnud opened the pull request:
<!--

Please prefix the title with the platform you are targetting:

Here are some examples of good titles:

- Desktop: Resolves #123: Added new setting to change font
- Mobile, Desktop: Fixes #456: Fixed config screen error
- All: Resolves #777: Made synchronisation faster

And here's an explanation of the title format:

- "Desktop" for the Windows/macOS/Linux app (Electron app)
- "Mobile" for the mobile app (or "Android" / "iOS" if the pull request only applies to one of the mobile platforms)
- "CLI" for the CLI app

If it's two platforms, separate them with commas - "Desktop, Mobile" or if it's for all platforms, prefix with "All".

If it's not related to any platform (such as a translation, change to the documentation, etc.), simply don't add a platform.

Then please append the issue that you've addressed or fixed. Use "Resolves #123" for new features or improvements and "Fixes #123" for bug fixes.

AND PLEASE READ THE GUIDE: https://github.com/laurent22/joplin/blob/dev/CONTRIBUTING.md

-->

I created an iOS Share Extension as discussed towards the end of https://discourse.joplinapp.org/t/ios-share-extension/51/27.

The extension is all native and works by copying the items passed to the Share Extension to a shared container that can be accessed by the main app. The extension then launches the main app via calling "joplin://shareData" and the main app reads the share data from this shared container and executes the same code Android was using to create a new note from the shared data. I also added handling for when the main app is already running to receive the shared data.

The share extension can receive text, a URL, as well up to 10 images, and 10 videos. This limit on images and videos is somewhat arbitrary, but can be easily changed.

Let me know if there are any issues with testing it. I changed the entitlements files to use app group "group.net.cozic.joplin" so it wouldn't include the one bound to my developer account. I think this new group identifier should get created automatically.

laurent22 commented:
Thanks for the pull request, it looks very good. I have a question about the pbxproj file. When upgrading React Native, previously we were provided with a new pbxproj, which we were supposed to diff with our own pbxproj file, however in practice that's pretty much impossible. As a result, I simply didn't update this file, which could cause issues, or start over the project from scratch by removing and re-adding all the dependencies, which of course is very time consuming.

In recent versions, React Native and package developers solved this nicely by relying on [autolinking](https://github.com/react-native-community/cli/blob/master/docs/autolinking.md) or in some cases by adding a line to the Podfile (as is done for react-native-vector-icons).

So what I'm wondering is, how does your extension handle upgrades? Is it possible to add this feature without making changes to pbxproj file, perhaps by supporting auto-linking or relying on the Podfile?

sirnacnud commented:
> Is it possible to add this feature without making changes to pbxproj file, perhaps by supporting auto-linking or relying on the Podfile?

We can include the source code files I added to the main app and the new share extension via auto-link/podspecs. But it's not possible to add this feature without modifying the pbxproj file, as the share extension target needs to be added, as well as configuring the main app and share extension to use the App Group Capabilities. I can change the PR to pull the source code in this way if that would be better.

In terms of how the share extension would be effected by upgrades, I don't think there is any automated way this could be handled. It seems like the choices are to use [rn-diff-purge](https://github.com/react-native-community/rn-diff-purge) or [Upgrade Helper](https://react-native-community.github.io/upgrade-helper/) to manually apply the changes to pbxproj file if there are conflicts during the upgrade. The other option would be be to regenerate the the project and then recreate the share extension target. Neither of these options are really ideal, but I'm happy to help out in the future if there are issues with the pbxproj file.

laurent22 commented:
Thanks for the detailed info and apologies for the late reply, things have been quite busy lately and I'm only starting to go over the pull requests.

What you've suggestion about pulling the source code via podspec would be great and easier to manage.

For the pbxproj, could you perhaps provide a doc file on what manual changes need to be made to install the extension? Basically the steps to follow in XCode. Then presumably once we know how to manually install it, we'll also know how to handle it during the RN upgrades.

sirnacnud commented:
I have moved most of the code into two podspecs. One is the react native module used in the app and the other is used in the Share Extension. I couldn't move all the code for the Share Extension into the podspec or else it would fail to run. I guess some compiled code needs to be part of the extension, separate from the linked pod library.

I also added a readme explaining how to create the Share Extension and configure the project. Let me know if any the steps aren't clear.

lukeswilliams commented:
Very keen for this to make it into the iOS app!

laurent22 commented on packages/app-mobile/root.tsx:
```diff
@@ -618,6 +618,12 @@ class AppComponent extends React.Component {
 		};
+
+		this.handleOpenURL_ = (event: any) => {
+			if (event.url == ShareExtension.shareURL) {
```
As I understand ShareExtension.shareURL will be an empty string in Android, but how about event.url? Is it also an empty string? Just trying to understand how it will work in Android.

roman-r-m replied:
I do not think this will be ever called on Android unless there's a corresponding change in manifest.
If I read it correctly, this listener will be called if user open a url like  `url://<....>`

sirnacnud replied:
Yes that's correct. For the iOS app, I added a `joplin://` URI scheme which launches the main app from the share extension. Nothing like this was added for Android, so `handleOpenURL` shouldn't be called.

laurent22 reviewed (commented):
Many thanks for the update and very useful Readme file. I only have one comment after that, and then we can merge.

laurent22 commented:
All good now, many thanks for your work @sirnacnud!

subzero355 commented:
Is it possible to share files to Joplin yet?
